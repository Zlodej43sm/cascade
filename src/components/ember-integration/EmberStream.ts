export interface IEmberStream<T> {
  subscribe(callback: (stream: IEmberStream<T>) => void): () => void;

  value(): T;

  setValue(value: T): void;
}

/**
 * Value wrapper solves data binding problem,
 * providing functionality to subscribe to value updates
 */
class EmberStream<T> implements IEmberStream<T> {
  private valueHolder?: T = undefined;

  private subscriptions = [] as ((stream: IEmberStream<T>) => void)[];

  constructor(value: T) {
    this.valueHolder = value;
  }

  /**
   * returns current value holding by wrapper
   */
  value(): T {
    return this.valueHolder!;
  }

  subscribe(callback: (stream: IEmberStream<T>) => void) {
    this.subscriptions.push(callback);
    return () => this.unsubscribe(callback);
  }

  unsubscribe(callback: (stream: IEmberStream<T>) => void) {
    this.subscriptions = this.subscriptions.filter((value) => value !== callback);
  }

  setValue(value: T) {
    this.valueHolder = value;
    this.subscriptions.forEach((subscription) => subscription(this));
  }
}

export default EmberStream;
