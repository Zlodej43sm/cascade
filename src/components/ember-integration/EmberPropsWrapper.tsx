import React, { ComponentType, PureComponent } from 'react';

import EmberStream from './EmberStream';

type Props = Record<string, EmberStream<any>>;

/**
 * component for the react injection to Ember
 */
const EmberPropsWrapper = <T extends unknown>(Comp: ComponentType<T>) => class Wrapper extends PureComponent<Props, T> {
  protected unsubscribers = [] as (() => void)[];

  constructor(props: Props) {
    super(props);
    const state: any = {};

    Object.entries(props).forEach(([key, stream]) => {
      const unsubscriber = stream.subscribe((s) => this.setState({ [key]: s.value() } as any));
      this.unsubscribers.push(unsubscriber);
      state[key] = stream.value();
    });

    this.state = state;
  }

  componentWillUnmount() {
    this.unsubscribers.forEach((unsubscriber) => unsubscriber());
  }

  render() {
    return <Comp {...this.state} />;
  }
};

export default EmberPropsWrapper;
