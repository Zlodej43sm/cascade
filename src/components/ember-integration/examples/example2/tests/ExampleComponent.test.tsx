import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import EmberStream from 'lib/components/ember-integration/EmberStream';
import EmberPropsWrapper from 'lib/components/ember-integration/EmberPropsWrapper';
import ExampleComponent, { MAGIC_NUMBER } from '../ExampleComponent';

const WrappedExampleComponent = EmberPropsWrapper(ExampleComponent);

describe('ExampleComponent unit tests', () => {
  const props = {
    name: new EmberStream('displayName-0'),
    data: new EmberStream({ Test: 1 }),
    callback: new EmberStream(jest.fn()),
  };

  it('Should match snapshot', () => {
    const { asFragment } = render(<WrappedExampleComponent {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should show contain name property', () => {
    const { getByTestId } = render(<WrappedExampleComponent {...props} />);

    expect(getByTestId('example-component').textContent).toContain(props.name.value());
  });

  it('Should update name property on changed', () => {
    const changedText = 'displayName-1';
    const { getByTestId } = render(<WrappedExampleComponent {...props} />);
    expect(getByTestId('example-component').textContent).not.toContain(changedText);

    props.name.setValue(changedText);
    expect(getByTestId('example-component').textContent).toContain(changedText);
  });

  it('Should call callback on button click', () => {
    const callbackMock = jest.fn();
    props.callback.setValue(callbackMock);
    const { getByText } = render(<WrappedExampleComponent {...props} />);
    expect(callbackMock).not.toBeCalled();

    fireEvent.click(getByText(/click me/i));

    expect(callbackMock).toBeCalled();
    expect(callbackMock).toBeCalledWith(MAGIC_NUMBER);
  });
});
