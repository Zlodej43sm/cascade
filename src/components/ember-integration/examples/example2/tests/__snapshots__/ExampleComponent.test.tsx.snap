// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleComponent unit tests Should match snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="example-component"
  >
    Hello from React to displayName-0 and data is: [object Object]!
    <button
      type="button"
    >
      Click me!
    </button>
  </div>
</DocumentFragment>
`;
